### PR TITLE
*: redesign `trace` statement with json output

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

There is a pending PR for improving tracing support https://github.com/pingcap/tidb/pull/8175

But I think the design is much too complex than necessary, it nearly roll up everything from ground and ... it have not update for some time.

Here is what I do, just collect data in the opentracing and output json data.

```
mysql> trace select count(1) from t, t1 where t.id = t1.id and t.id > 0 and t1.id < 666 order by t1.value limit 5;
```

The result looks like this:

```
[{"ID":{"Trace":"52036b29f98ab854","Span":"7afa53cb2733c854","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c2Vzc2lvbi5nZXRUeG5GdXR1cmU="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODU5MTk1NzYrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODU5NDQxNTQrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":[{"ID":{"Trace":"52036b29f98ab854","Span":"2dda78255f0b22dd","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c2Vzc2lvbi5QYXJzZVNRTA=="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODYwNDA1OSswODowMA=="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODYxNjk4NzMrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"135cb102a25ead32","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"ZXhlY3V0b3IuQ29tcGlsZQ=="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODYyMzc0NzIrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODcxMTE0NjkrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"7e493f542406d540","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c2Vzc2lvbi5Db21taXRUeG4="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc1MjI2ODkrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc1MzkxOTgrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"2c79ef314f1ac6de","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c2Vzc2lvbi5ydW5TdG10"},{"Key":"_schema:name","Value":null},{"Key":"Msg","Value":"eyJzcWwiOiJzZWxlY3QgY291bnQoMSkgZnJvbSB0LCB0MSB3aGVyZSB0LmlkID0gdDEuaWQgYW5kIHQuaWQgXHUwMDNlIDAgYW5kIHQxLmlkIFx1MDAzYyA2NjYifQ=="},{"Key":"Time","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODcxNzEzNDgrMDg6MDA="},{"Key":"_schema:log","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODcxNjA3NDgrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc1NzY4MTErMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"3a062efbf24e10ec","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c2Vzc2lvbi5FeGVjdXRl"},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODU5MTU0NDgrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc3MzExMjYrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"376b4b8df3739409","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"dGFibGVSZWFkZXIuTmV4dA=="},{"Key":"_schema:name","Value":null},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40Mjk4NDcwMTErMDg6MDA="},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc4ODM0MTkrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"3651443c69f0ccfa","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"dGFibGVSZWFkZXIuTmV4dA=="},{"Key":"_schema:name","Value":null},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcwMzM0NzYrMDg6MDA="},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc4ODI5NDQrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"3dc31232bfa6a458","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c3RyZWFtQWdnLk5leHQ="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc3NzgxMDMrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcwODY2MzUrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"40df8c16230c129f","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"ZXhlY3V0b3IuTmV4dA=="},{"Key":"_schema:name","Value":null},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcwOTYxODgrMDg6MDA="},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc3NzQ4NjErMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"75a70b5f168d0356","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"c3RyZWFtQWdnLk5leHQ="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcxMDc4NjIrMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcxMDgxNzcrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"73585aed20b6a99d","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"ZXhlY3V0b3IuTmV4dA=="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcxMDU0MzErMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcxMTE4MTkrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null},{"ID":{"Trace":"52036b29f98ab854","Span":"19885f4eda8dc1ac","Parent":"6d3a3f912a874491"},"Annotations":[{"Key":"Name","Value":"ZXhlY3V0b3IuTmV4dA=="},{"Key":"_schema:name","Value":null},{"Key":"Span.Start","Value":"MjAxOC0xMS0xOVQyMTowMzo0OC44ODc3NzAzNzErMDg6MDA="},{"Key":"Span.End","Value":"MjAxOC0xMS0xOVQyMTowMzo0OS40ODcxMTQ5NzUrMDg6MDA="},{"Key":"_schema:Timespan","Value":null}],"Sub":null}]}] 
```

Visualize those data in the web tools is another story.

image

### What is changed and how it works?

"trace select xxx from t where ..." give a json output
The json data could be visualized on the web UI by tools

We can embed the web UI in TiDB later, extract what we need.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

```
go get -u -v http://github.com/tiancaiamao/appdash/cmd
appdash demo
```
Visit http://localhost:8700/traces, there is a `input json` button, put the  json data of the trace output and view it.

Code changes

Set trace format to "json" and it's now the default one, "row" format is still kept

Related changes

 - Need to update the documentation

@lysu  @jackysp @shenli

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8357)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
@lysu @jackysp  @shenli

File: executor/adapter.go [hunk 1]
```diff
@@ -94,6 +95,11 @@ func schema2ResultFields(schema *expression.Schema, defaultDB string) (rfs []*as
 // next query.
 // If stmt is not nil and chunk with some rows inside, we simply update last query found rows by the number of row in chunk.
 func (a *recordSet) Next(ctx context.Context, chk *chunk.Chunk) error {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("executor.Next", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+	}
+
```
Comment (lysu) [inline executor/adapter.go]: How about:

```suggestion
var noop = func() {}

func trace(ctx context.Context, operationName string) func() {
	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
		span1 := span.Tracer().StartSpan(operationName, opentracing.ChildOf(span.Context()))
		return span1.Finish
	}
	return noop
}

func (a *recordSet) Next(ctx context.Context, chk *chunk.Chunk) error {
     defer trace(ctx, "executor.Next")
```
does we take care about `defer` performance?

**Comment by zhexuany** [comment]
@tiancaiamao Can we hold this PR for a few days? I will try to merge all trace related PR in nearly future.

**Comment by tiancaiamao** [comment]
This PR doesn't conflict with other work.
The `trace format = row` is still available and you can fulfill that. @zhexuany

**Comment by zhexuany** [comment]
Gotcha, I already approved this PR. Thanks.

**Comment by tiancaiamao** [comment]
Need parser changes https://github.com/pingcap/parser/pull/46

**Pull request merged**